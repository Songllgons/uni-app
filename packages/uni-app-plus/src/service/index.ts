import * as uni from './api'
import { UniServiceJSBridge } from './bridge'
import { registerApp as __registerApp, getApp } from './framework/app'
import {
  definePage as __definePage,
  registerPage as __registerPage,
  getCurrentPages,
} from './framework/page'

export default {
  uni,
  getApp,
  getCurrentPages,
  __definePage,
  __registerApp,
  __registerPage,
  UniServiceJSBridge,
}
