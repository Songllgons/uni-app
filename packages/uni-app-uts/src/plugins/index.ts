import { init as initAndroidPlugins } from './android'
import { init as initIOSPlugins } from './ios'

export function initAndroid() {
  import('./errorReporting')
  return initAndroidPlugins()
}

export function initIOS() {
  import('./errorReporting')
  return initIOSPlugins()
}
