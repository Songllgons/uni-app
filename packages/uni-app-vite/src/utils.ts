import path from 'path'
export const templateDir = path.resolve(__dirname, '../lib/template/')
export function nvueOutDir() {
  return path.join(process.env.UNI_OUTPUT_DIR, '../.nvue')
}
