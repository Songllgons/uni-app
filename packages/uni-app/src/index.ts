export * from './ssr'
export * from './api'
export * from './hbx'
export * from './utils'
export * from './apiLifecycle'
export * from './vue'
