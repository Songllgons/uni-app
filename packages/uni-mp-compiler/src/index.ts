import { extend } from '@vue/shared'
import {
  type CodegenResult,
  type ParserOptions,
  type RootNode,
  baseParse,
} from '@vue/compiler-core'

import { baseCompile } from './compile'
import { parserOptions } from './parserOptions'
import type { CompilerOptions } from './options'
export type { CompilerOptions } from './options'
export { findProp } from '@vue/compiler-core'
export type {
  DirectiveNode,
  NodeTransform,
  DirectiveTransform,
  TransformContext,
  SimpleExpressionNode,
} from '@vue/compiler-core'

export { genExpr } from './codegen'
export { rewriteExpression } from './transforms/utils'
export { isForElementNode } from './transforms/vFor'
export { transformOn } from './transforms/vOn'
export { transformModel } from './transforms/vModel'
export * from './runtimeHelpers'

export function parse(template: string, options: ParserOptions = {}): RootNode {
  return baseParse(template, extend({}, parserOptions, options))
}

export function compile(
  template: string,
  options: CompilerOptions = {}
): CodegenResult {
  return baseCompile(
    template,
    extend({}, parserOptions, options, {
      directiveTransforms: extend({}, options.directiveTransforms || {}),
    })
  )
}
