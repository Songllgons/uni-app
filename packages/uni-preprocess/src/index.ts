export { preprocess } from './preprocess'
export type {
  PreprocessOptions,
  ProcessContext,
  SourceMapOptions,
} from './preprocess'
