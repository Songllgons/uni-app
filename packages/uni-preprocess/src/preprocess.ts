import MagicString, { type SourceMapOptions } from 'magic-string'
import { matchRecursive } from 'xregexp'

export type { SourceMapOptions } from 'magic-string'

export type ProcessContext = Record<string, any>
export interface PreprocessOptions {
  type?: 'js' | 'html'
  context: ProcessContext
  sourceMap?: SourceMapOptions
}

export function preprocess(source: string, options: PreprocessOptions) {
  let isInPreprocessor = (_offset: number) => false
  if (!source.includes('#endif')) {
    return {
      code: source,
      map: options.sourceMap
        ? new MagicString(source).generateMap(options.sourceMap)
        : null,
      isInPreprocessor,
    }
  }

  const context = options.context || {}
  const s = new MagicString(source)

  // [[start,end]]
  const ranges: [number, number][] = []

  isInPreprocessor = (offset: number) => {
    for (let i = 0; i < ranges.length; i++) {
      const range = ranges[i]
      if (range[0] <= offset && offset < range[1]) {
        return true
      }
    }
    return false
  }

  function preprocessByType(type: PreprocessType) {
    replaceRecursive(
      type,
      source,
      s,
      (s, startMatches, endMatches, include, recurse) => {
        // I need to recurse first, so I don't catch "inner" else-directives
        recurse(include.value, include.start)

        const variant = startMatches.value[1]
        const test = (startMatches.value[2] || '').trim()
        switch (variant) {
          case 'ifdef':
            if (testPasses(test, context)) {
              s.remove(startMatches.start, startMatches.end)
              s.remove(endMatches.start, endMatches.end)
            } else {
              s.remove(startMatches.start, endMatches.end)
            }
            ranges.push([startMatches.start, endMatches.end])
            return
          case 'ifndef':
            if (!testPasses(test, context)) {
              s.remove(startMatches.start, startMatches.end)
              s.remove(endMatches.start, endMatches.end)
            } else {
              s.remove(startMatches.start, endMatches.end)
            }
            ranges.push([startMatches.start, endMatches.end])
            return
          default:
            throw new Error('Unknown if variant ' + variant + '.')
        }
      }
    )
  }
  const type = options.type || 'auto'
  if (type === 'auto' || type === 'js') {
    preprocessByType(TYPES.js)
  }
  if (type === 'auto' || type === 'html') {
    preprocessByType(TYPES.html)
  }
  return {
    code: s.toString(),
    map: options.sourceMap ? s.generateMap(options.sourceMap) : null,
    isInPreprocessor,
  }
}

type Processor = (
  s: MagicString,
  startMatches: MatchGroupLeft,
  endMatches: MatchGroupRight,
  include: MatchGroupMatch,
  recurse: (content: string, startOffset: number) => void
) => void

const startPattern =
  '[ \t]*(?://|/\\*)[ \t]*#(ifndef|ifdef)[ \t]+([^\n*]*)(?:\\*(?:\\*|/))?(?:[ \t]*\n+)?'
const endPattern =
  '[ \t]*(?://|/\\*)[ \t]*#endif[ \t]*(?:\\*(?:\\*|/))?(?:[ \t]*\n)?'

const startPatternHTML =
  '[ \t]*<!--[ \t]*#(ifndef|ifdef|if)[ \t]+(.*?)[ \t]*(?:-->|!>)(?:[ \t]*\n+)?'
const endPatternHTML =
  '[ \t]*<!(?:--)?[ \t]*#endif[ \t]*(?:-->|!>)(?:[ \t]*\n)?'

interface PreprocessType {
  start: {
    pattern: string
    regex: RegExp
  }
  end: {
    pattern: string
    regex: RegExp
  }
}

const TYPES: Record<string, PreprocessType> = {
  js: {
    start: {
      pattern: startPattern,
      regex: new RegExp(startPattern, 'mi'),
    },
    end: {
      pattern: endPattern,
      regex: new RegExp(endPattern, 'mi'),
    },
  },
  html: {
    start: {
      pattern: startPatternHTML,
      regex: new RegExp(startPatternHTML, 'mi'),
    },
    end: {
      pattern: endPatternHTML,
      regex: new RegExp(endPatternHTML, 'mi'),
    },
  },
}

interface MatchGroupItem {
  start: number
  end: number
}
interface MatchGroupMatch extends MatchGroupItem {
  value: string
}
interface MatchGroupLeft extends MatchGroupItem {
  value: RegExpExecArray
}

type MatchGroupRight = MatchGroupLeft

function replaceRecursive(
  type: PreprocessType,
  source: string,
  s: MagicString,
  processor: Processor
) {
  function matchReplacePass(content: string, startOffset: number = 0) {
    const matches = matchRecursive(
      content,
      type.start.pattern,
      type.end.pattern,
      'gmi',
      {
        valueNames: ['between', 'left', 'match', 'right'],
      }
    )

    let left: MatchGroupLeft | null = null
    let match: MatchGroupMatch | null = null
    let right: MatchGroupRight | null = null
    matches.forEach(function ({ name, value, start, end }) {
      start = start + startOffset
      end = end + startOffset
      switch (name) {
        case 'between':
          break
        case 'left':
          left = {
            start,
            end,
            value: type.start.regex.exec(value)!,
          }
          break
        case 'match':
          match = {
            start,
            end,
            value,
          }
          break
        case 'right':
          right = {
            start,
            end,
            value: type.end.regex.exec(value)!,
          }
          if (left && match && right) {
            processor(s, left, right, match, matchReplacePass)
            left = null
            match = null
            right = null
          }
          break
      }
    })
  }

  matchReplacePass(source)
}

function getTestTemplate(test: string) {
  test = test || 'true'
  test = test.trim()

  // force single equals replacement
  // 不替换，会影响 >= 等判断
  // test = test.replace(/([^=!])=([^=])/g, '$1==$2');
  test = test.replace(/-/g, '_')
  return new Function(
    'context',
    'with (context||{}){ return ( ' + test + ' ); }'
  )
}

function testPasses(test: string, context: ProcessContext) {
  var testFn = getTestTemplate(test)
  try {
    return testFn(context)
  } catch (e) {}
  return false
}
