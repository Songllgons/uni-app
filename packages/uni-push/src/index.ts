import GtPush from '../lib/gtpush-min'
import { initPushNotification } from './route'
import { initBroadcastChannel, postPushMessage } from './broadcastChannel'

// if (process.env.UNI_PUSH_DEBUG) {
//   GtPush.setDebugMode(true)
// }

// @ts-expect-error
uni.invokePushCallback({
  type: 'enabled',
})

const appid = process.env.UNI_APP_ID!
if (!appid) {
  Promise.resolve().then(() => {
    // @ts-expect-error
    uni.invokePushCallback({
      type: 'clientId',
      cid: '',
      errMsg: 'manifest.json->appid is required',
    })
  })
} else {
  // #ifdef APP
  initPushNotification()
  // #endif
  // #ifdef H5
  initBroadcastChannel(GtPush)
  // #endif
  // #ifdef MP || APP
  if (typeof uni.onAppShow === 'function') {
    uni.onAppShow(() => {
      GtPush.enableSocket(true)
    })
  }
  // #endif
  GtPush.init({
    appid,
    onError: (res) => {
      console.error(res.error)
      const data = {
        type: 'clientId',
        cid: '',
        errMsg: res.error,
      }
      // @ts-expect-error
      uni.invokePushCallback(data)
      // #ifdef H5
      postPushMessage(data)
      // #endif
    },
    onClientId: (res) => {
      const data = {
        type: 'clientId',
        cid: res.cid,
      }
      // @ts-expect-error
      uni.invokePushCallback(data)
      // #ifdef H5
      postPushMessage(data)
      // #endif
    },
    onlineState: (res) => {
      const data = {
        type: 'lineState',
        online: res.online,
      }
      // @ts-expect-error
      uni.invokePushCallback(data)
      // #ifdef H5
      postPushMessage(data)
      // #endif
    },
    onPushMsg: (res) => {
      const data = {
        type: 'pushMsg',
        message: res.message,
      }
      // @ts-expect-error
      uni.invokePushCallback(data)
      // #ifdef H5
      postPushMessage(data)
      // #endif
    },
  })
  // 仅在 jssdk 中监听
  // #ifdef APP
  uni.onPushMessage((res) => {
    if (
      res.type === 'receive' &&
      res.data &&
      (res.data as any).force_notification
    ) {
      // 创建通知栏
      uni.createPushMessage(res.data)
      // 阻止其他监听器继续监听
      ;(res as any).stopped = true
    }
  })
  // #endif
}
