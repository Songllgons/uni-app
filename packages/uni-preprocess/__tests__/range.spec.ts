// import { writeFileSync } from 'node:fs'
// import { resolve } from 'node:path'
import { preprocess } from '../src/index'

describe('preprocess', () => {
  test('isInPreprocessor', () => {
    const code = `
.test {
    color:red
}
// #ifdef B
.test {
    color:blue
}
// #ifdef B1
.test {
    color:yellow
}
// #endif
.test {
    color:black
}
// #endif
.test {
    color:green
}
    `
    const { isInPreprocessor } = preprocess(code, {
      context: { B: true },
    })
    const result: [number, string, boolean][] = []
    for (let i = 0; i < code.length; i++) {
      result.push([i, code[i], isInPreprocessor(i)])
    }
    expect(result).toMatchSnapshot()
  })
})
