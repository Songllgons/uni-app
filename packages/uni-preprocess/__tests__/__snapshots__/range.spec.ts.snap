// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preprocess isInPreprocessor 1`] = `
[
  [
    0,
    "
",
    false,
  ],
  [
    1,
    ".",
    false,
  ],
  [
    2,
    "t",
    false,
  ],
  [
    3,
    "e",
    false,
  ],
  [
    4,
    "s",
    false,
  ],
  [
    5,
    "t",
    false,
  ],
  [
    6,
    " ",
    false,
  ],
  [
    7,
    "{",
    false,
  ],
  [
    8,
    "
",
    false,
  ],
  [
    9,
    " ",
    false,
  ],
  [
    10,
    " ",
    false,
  ],
  [
    11,
    " ",
    false,
  ],
  [
    12,
    " ",
    false,
  ],
  [
    13,
    "c",
    false,
  ],
  [
    14,
    "o",
    false,
  ],
  [
    15,
    "l",
    false,
  ],
  [
    16,
    "o",
    false,
  ],
  [
    17,
    "r",
    false,
  ],
  [
    18,
    ":",
    false,
  ],
  [
    19,
    "r",
    false,
  ],
  [
    20,
    "e",
    false,
  ],
  [
    21,
    "d",
    false,
  ],
  [
    22,
    "
",
    false,
  ],
  [
    23,
    "}",
    false,
  ],
  [
    24,
    "
",
    false,
  ],
  [
    25,
    "/",
    true,
  ],
  [
    26,
    "/",
    true,
  ],
  [
    27,
    " ",
    true,
  ],
  [
    28,
    "#",
    true,
  ],
  [
    29,
    "i",
    true,
  ],
  [
    30,
    "f",
    true,
  ],
  [
    31,
    "d",
    true,
  ],
  [
    32,
    "e",
    true,
  ],
  [
    33,
    "f",
    true,
  ],
  [
    34,
    " ",
    true,
  ],
  [
    35,
    "B",
    true,
  ],
  [
    36,
    "
",
    true,
  ],
  [
    37,
    ".",
    true,
  ],
  [
    38,
    "t",
    true,
  ],
  [
    39,
    "e",
    true,
  ],
  [
    40,
    "s",
    true,
  ],
  [
    41,
    "t",
    true,
  ],
  [
    42,
    " ",
    true,
  ],
  [
    43,
    "{",
    true,
  ],
  [
    44,
    "
",
    true,
  ],
  [
    45,
    " ",
    true,
  ],
  [
    46,
    " ",
    true,
  ],
  [
    47,
    " ",
    true,
  ],
  [
    48,
    " ",
    true,
  ],
  [
    49,
    "c",
    true,
  ],
  [
    50,
    "o",
    true,
  ],
  [
    51,
    "l",
    true,
  ],
  [
    52,
    "o",
    true,
  ],
  [
    53,
    "r",
    true,
  ],
  [
    54,
    ":",
    true,
  ],
  [
    55,
    "b",
    true,
  ],
  [
    56,
    "l",
    true,
  ],
  [
    57,
    "u",
    true,
  ],
  [
    58,
    "e",
    true,
  ],
  [
    59,
    "
",
    true,
  ],
  [
    60,
    "}",
    true,
  ],
  [
    61,
    "
",
    true,
  ],
  [
    62,
    "/",
    true,
  ],
  [
    63,
    "/",
    true,
  ],
  [
    64,
    " ",
    true,
  ],
  [
    65,
    "#",
    true,
  ],
  [
    66,
    "i",
    true,
  ],
  [
    67,
    "f",
    true,
  ],
  [
    68,
    "d",
    true,
  ],
  [
    69,
    "e",
    true,
  ],
  [
    70,
    "f",
    true,
  ],
  [
    71,
    " ",
    true,
  ],
  [
    72,
    "B",
    true,
  ],
  [
    73,
    "1",
    true,
  ],
  [
    74,
    "
",
    true,
  ],
  [
    75,
    ".",
    true,
  ],
  [
    76,
    "t",
    true,
  ],
  [
    77,
    "e",
    true,
  ],
  [
    78,
    "s",
    true,
  ],
  [
    79,
    "t",
    true,
  ],
  [
    80,
    " ",
    true,
  ],
  [
    81,
    "{",
    true,
  ],
  [
    82,
    "
",
    true,
  ],
  [
    83,
    " ",
    true,
  ],
  [
    84,
    " ",
    true,
  ],
  [
    85,
    " ",
    true,
  ],
  [
    86,
    " ",
    true,
  ],
  [
    87,
    "c",
    true,
  ],
  [
    88,
    "o",
    true,
  ],
  [
    89,
    "l",
    true,
  ],
  [
    90,
    "o",
    true,
  ],
  [
    91,
    "r",
    true,
  ],
  [
    92,
    ":",
    true,
  ],
  [
    93,
    "y",
    true,
  ],
  [
    94,
    "e",
    true,
  ],
  [
    95,
    "l",
    true,
  ],
  [
    96,
    "l",
    true,
  ],
  [
    97,
    "o",
    true,
  ],
  [
    98,
    "w",
    true,
  ],
  [
    99,
    "
",
    true,
  ],
  [
    100,
    "}",
    true,
  ],
  [
    101,
    "
",
    true,
  ],
  [
    102,
    "/",
    true,
  ],
  [
    103,
    "/",
    true,
  ],
  [
    104,
    " ",
    true,
  ],
  [
    105,
    "#",
    true,
  ],
  [
    106,
    "e",
    true,
  ],
  [
    107,
    "n",
    true,
  ],
  [
    108,
    "d",
    true,
  ],
  [
    109,
    "i",
    true,
  ],
  [
    110,
    "f",
    true,
  ],
  [
    111,
    "
",
    true,
  ],
  [
    112,
    ".",
    true,
  ],
  [
    113,
    "t",
    true,
  ],
  [
    114,
    "e",
    true,
  ],
  [
    115,
    "s",
    true,
  ],
  [
    116,
    "t",
    true,
  ],
  [
    117,
    " ",
    true,
  ],
  [
    118,
    "{",
    true,
  ],
  [
    119,
    "
",
    true,
  ],
  [
    120,
    " ",
    true,
  ],
  [
    121,
    " ",
    true,
  ],
  [
    122,
    " ",
    true,
  ],
  [
    123,
    " ",
    true,
  ],
  [
    124,
    "c",
    true,
  ],
  [
    125,
    "o",
    true,
  ],
  [
    126,
    "l",
    true,
  ],
  [
    127,
    "o",
    true,
  ],
  [
    128,
    "r",
    true,
  ],
  [
    129,
    ":",
    true,
  ],
  [
    130,
    "b",
    true,
  ],
  [
    131,
    "l",
    true,
  ],
  [
    132,
    "a",
    true,
  ],
  [
    133,
    "c",
    true,
  ],
  [
    134,
    "k",
    true,
  ],
  [
    135,
    "
",
    true,
  ],
  [
    136,
    "}",
    true,
  ],
  [
    137,
    "
",
    true,
  ],
  [
    138,
    "/",
    true,
  ],
  [
    139,
    "/",
    true,
  ],
  [
    140,
    " ",
    true,
  ],
  [
    141,
    "#",
    true,
  ],
  [
    142,
    "e",
    true,
  ],
  [
    143,
    "n",
    true,
  ],
  [
    144,
    "d",
    true,
  ],
  [
    145,
    "i",
    true,
  ],
  [
    146,
    "f",
    true,
  ],
  [
    147,
    "
",
    true,
  ],
  [
    148,
    ".",
    false,
  ],
  [
    149,
    "t",
    false,
  ],
  [
    150,
    "e",
    false,
  ],
  [
    151,
    "s",
    false,
  ],
  [
    152,
    "t",
    false,
  ],
  [
    153,
    " ",
    false,
  ],
  [
    154,
    "{",
    false,
  ],
  [
    155,
    "
",
    false,
  ],
  [
    156,
    " ",
    false,
  ],
  [
    157,
    " ",
    false,
  ],
  [
    158,
    " ",
    false,
  ],
  [
    159,
    " ",
    false,
  ],
  [
    160,
    "c",
    false,
  ],
  [
    161,
    "o",
    false,
  ],
  [
    162,
    "l",
    false,
  ],
  [
    163,
    "o",
    false,
  ],
  [
    164,
    "r",
    false,
  ],
  [
    165,
    ":",
    false,
  ],
  [
    166,
    "g",
    false,
  ],
  [
    167,
    "r",
    false,
  ],
  [
    168,
    "e",
    false,
  ],
  [
    169,
    "e",
    false,
  ],
  [
    170,
    "n",
    false,
  ],
  [
    171,
    "
",
    false,
  ],
  [
    172,
    "}",
    false,
  ],
  [
    173,
    "
",
    false,
  ],
  [
    174,
    " ",
    false,
  ],
  [
    175,
    " ",
    false,
  ],
  [
    176,
    " ",
    false,
  ],
  [
    177,
    " ",
    false,
  ],
]
`;
